### PR TITLE
Feature remove forced unwrapping

### PR DESCRIPTION
**Pull request by mrfishball**

### CODE REVIEW

**Comment by ericmeyer** [comment]
What story/Trello card is this related to?

File: Sources/SwerverCore/Router/Router.swift [hunk 1]
```diff
@@ -12,10 +12,11 @@ public class Router {
         populateRoutes()
     }
 
-    public func process(request: HttpRequest) -> String {
+    public func process(request: HttpRequest?) -> String {
 
-        let targetURL = request.getUrl()
-        let targetMethod = request.getMethod()
+        guard let targetURL = request?.getUrl(), let targetMethod = request?.getMethod() else {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Router.swift]: Typically, it would be conventional to unwrap the `request` itself once, and not for each attribute on the request.

e.g.
```swift
func route(request: String?) -> String {
    if let request = request {
        return "Valid Request: \(request)"
    } else {
        return "Invalid Request"
    }
}

print(route(request: "Good")) // "Valid Request: Good"
print(route(request: nil)) // "Invalid Request
```

File: Sources/SwerverCore/Router/Router.swift [hunk 2]
```diff
@@ -33,11 +34,13 @@ public class Router {
         return responseHeaderFormatter.format(httpResponse: notFoundAction.dispatch())
     }
 
-    private func isOptionsRequest(request: HttpRequest) -> Bool {
-        return request.getMethod() == RequestMethod.options
+    private func isOptionsRequest(request: HttpRequest?) -> Bool {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Router.swift]: In theory, this should never be called if the request is nil, therefore you can probably keep this as `HttpRequest` instead of `HttpRequest?`. This is related to my other comment about unwrapping `request`

File: Sources/SwerverCore/Router/Routes.swift [hunk 3]
```diff
@@ -22,7 +22,7 @@ public class Routes {
     }
     
     public func fetchAllActions(url: URL) -> [RequestMethod:HttpAction] {
-        return allRoutes[url]!
+        return allRoutes[url] ?? [:]
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Routes.swift]: This is a cool use of `??`. What exactly does `??` do?

File: Sources/SwerverCore/Server/ClientRequestHandler.swift [hunk 4]
```diff
@@ -10,18 +10,18 @@ public class ClientRequestHandler {
     
     public func handle(client: Socket) throws {
         let httpRequest = try parseRequest(clientSocket: client)
-        try client.write(from: httpRouteProcessor.process(request: httpRequest))
+        try client.write(from: httpRouteProcessor.process(request: (httpRequest)))
```
Comment (ericmeyer) [inline Sources/SwerverCore/Server/ClientRequestHandler.swift]: Why did you need to add parens here?

File: Sources/SwerverCore/Server/ClientRequestHandler.swift [hunk 5]
```diff
@@ -10,18 +10,18 @@ public class ClientRequestHandler {
...
     }
     
-    private func parseRequest(clientSocket: Socket) throws -> HttpRequest {
-        var readData: Data? = Data(capacity: Swerver.bufferSize)
-        let _ = try clientSocket.read(into: &readData!)
-        guard let parsedData = String(data: readData!, encoding: .utf8) else {
-            readData!.count = 0
+    private func parseRequest(clientSocket: Socket) throws -> HttpRequest? {
+        var readData: Data = Data(capacity: Swerver.bufferSize)
+        let _ = try clientSocket.read(into: &readData)
```
Comment (ericmeyer) [inline Sources/SwerverCore/Server/ClientRequestHandler.swift]: Can `let _ = ` be removed here?

File: Sources/SwerverCore/Server/ClientRequestHandler.swift [hunk 6]
```diff
@@ -10,18 +10,18 @@ public class ClientRequestHandler {
...
+        guard let parsedData = String(data: readData, encoding: .utf8) else {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Server/ClientRequestHandler.swift]: Given that there are two distinct branches here, I think an `if let` with an explicit `else` branch would be easier to read than a `guard`

File: Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift [hunk 7]
```diff
@@ -11,10 +11,19 @@ class HttpRequestParserSpec: QuickSpec {
             it("returns an HttpRequest Object") {
                 let requestData = "GET / HTTP/1.1"
                 let parsedRequest = httpRequestParser.parse(request: requestData)
-                expect(parsedRequest.getHttpVersion()).to(equal(HttpVersion.current))
-                expect(parsedRequest.getUrl()).to(equal(URL(string: "/")))
-                expect(parsedRequest.getMethod()).to(equal(RequestMethod.get))
+                expect(parsedRequest?.getHttpVersion()).to(equal(HttpVersion.current))
```
Comment (ericmeyer) [inline Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift]: This should be split into multiple tests. One for each attribute

File: Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift [hunk 8]
```diff
@@ -11,10 +11,19 @@ class HttpRequestParserSpec: QuickSpec {
...
+                expect(parsedRequest?.getUrl()).to(equal(URL(string: "/")))
+                expect(parsedRequest?.getMethod()).to(equal(RequestMethod.get))
             }
+            
+            context("when the request string is invalid") {
+                it("returns nil") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift:25]: Great test.

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift [hunk 9]
```diff
@@ -9,10 +9,27 @@ class RouterSpec: QuickSpec {
             var router = Router()
```
Comment (ericmeyer) [inline Tests/SwerverTests/RouterSpec/RouterSpec.swift:9]: This can be part of a different refactoring, but I would avoid creating objects outside the scope of a test. The danger here is that the same instance will persist from test to test, resulting in the effects of one test persisting to the next test run. This can cause issues that are very hard to debug.

In general, tests should be completely isolated from each other, and a given test should not depend on another test or tests.

This change is definitely out of scope for this pull request.

**Review by ericmeyer — CHANGES_REQUESTED** [review]

**Comment by ericmeyer** [comment]
Another thing to note: You are definitely going to have merge conflicts to resolve if/when one of the pull requests gets merged in.

I recommend merging in the larger PR first, since resolving merge conflicts in the smaller branch will be easier than the doing it the other order.

File: Sources/SwerverCore/Router/Router.swift (hunk 2 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Router.swift]: this function is actually removed in the feature-refactoring branch as it's no longer needed

File: Sources/SwerverCore/Router/Router.swift (hunk 1 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Router.swift]: will fix that

File: Sources/SwerverCore/Router/Routes.swift (hunk 3 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Routes.swift]: ?? is used to provide a fallback if the predecessor is nil, in this case, it'll return an empty dictionary

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 6 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: will look into that

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 5 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: will look into that

File: Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift (hunk 7 above)
Comment (mrfishball) [reply Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift]: will look into that

File: Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift (hunk 8 above)
Comment (mrfishball) [reply Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift:25]: 😃

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 9 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:9]: ok

**Review by cmvandrevala — APPROVED** [review]

File: Sources/SwerverCore/Router/Router.swift (hunk 1 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Router.swift]: fixed

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 5 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: if I remove it, the compiler will throw a warning. I believe .read returns a Data object

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 6 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: fixed

File: Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift (hunk 7 above)
Comment (mrfishball) [reply Tests/SwerverTests/RequestSpec/HttpRequestParserSpec.swift]: fixed

File: Sources/SwerverCore/Router/Routes.swift (hunk 3 above)
Comment (ericmeyer) [reply Sources/SwerverCore/Router/Routes.swift]: Very cool.

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 5 above)
Comment (ericmeyer) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: :+1:
Weird code is slightly preferable to a compiler warning.

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 9 above)
Comment (ericmeyer) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:9]: I would consider how to keep track of all the comments that are "pushed" to future commits. It might be something as simple as creating cards in Trello and shoving them in the backlog.

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 4 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: fixed

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 9 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:9]: will track those using trello cards

File: Sources/SwerverCore/Server/ClientRequestHandler.swift (hunk 5 above)
Comment (mrfishball) [reply Sources/SwerverCore/Server/ClientRequestHandler.swift]: 👍

File: Sources/SwerverCore/Router/Router.swift [hunk 10]
```diff
@@ -10,27 +10,29 @@ public class Router {
...
+                
+                guard let action = targetRoute[targetMethod] else {
+                    return notFoundAction.execute()
+                }
+                return action.execute()
             }
-            return responseHeaderFormatter.format(httpResponse: action.execute())
         }
-        return responseHeaderFormatter.format(httpResponse: notFoundAction.execute())
+        return notFoundAction.execute()
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Router.swift:29]: Nitpick: I would put this in an `else` branch. It makes it slightly more explicit that these branches are mutually exclusive.

File: Tests/SwerverTests/RequestSpec/HttpRequestSpec.swift [hunk 11]
```diff
@@ -7,11 +7,17 @@ class HttpRequestSpec: QuickSpec {
 
     override func spec() {
         describe("HttpRequest") {
-            var httpRequest = HttpRequest(method: RequestMethod.get, url: URL(string: "/")!)
+            guard let url = URL(string: "/") else {
+                return
+            }
+            var httpRequest = HttpRequest(method: RequestMethod.get, url: url)
```
Comment (ericmeyer) [inline Tests/SwerverTests/RequestSpec/HttpRequestSpec.swift:13]: What are your thoughts on making the `HttpRequest` constructor private, and introducing an `HttpRequestBuilder`?

#futurepullrequest :)

**Review by ericmeyer — APPROVED** [review]

**Comment by ericmeyer** [comment]
:shipit: !

File: Sources/SwerverCore/Router/Router.swift (hunk 10 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Router.swift:29]: will fixed that

File: Tests/SwerverTests/RequestSpec/HttpRequestSpec.swift (hunk 11 above)
Comment (mrfishball) [reply Tests/SwerverTests/RequestSpec/HttpRequestSpec.swift:13]: that thought came to me many times but for the sake of not over-engineering, I guess I should leave it as it is until that time comes

Comment (mrfishball) [reply Tests/SwerverTests/RequestSpec/HttpRequestSpec.swift:13]: will also track it in trello
